File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque


def solution(numbers, target):
    answer = 0
    queue = deque()
    
    queue.append((numbers[0], 0))
    queue.append((-numbers[0],0))
    
    
    while queue:
        nowNum, depth = queue.popleft()
        if(depth < len(numbers) - 1):
            queue.append((nowNum + numbers[depth+1], depth+1))
            queue.append((nowNum - numbers[depth+1], depth+1))
        else:
            if nowNum == target:
                answer +=1
    return answer

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer;
    public int solution(int[] numbers, int target) {
        answer = 0;
        
        backTrack(1, numbers, target, numbers[0]);
        backTrack(1, numbers, target, -numbers[0]);
        
        
        return answer;
    }
    
    
    public void backTrack(int idx, int[] numbers, int target, int now){
        if(idx == numbers.length){
            if(now == target){
                answer++;
            }
            return;
        }
        backTrack(idx+1, numbers, target, now + numbers[idx]);
        backTrack(idx+1, numbers, target, now - numbers[idx]);
        
    }
}